pkg/binding/bindingpolicy.go

```diff
@@ -81,24 +76,45 @@ func (c *Controller) handleBindingPolicy(ctx context.Context, obj runtime.Object
 		// note bindingpolicy in resolver in case it isn't associated with
 		// any resolution
 		c.bindingPolicyResolver.NoteBindingPolicy(bindingPolicy)
+
+		if bindingPolicy.Spec.WantSingletonReportedState {
+			// if the bindingpolicy requires a singleton status, then we should only
+			// have one destination
+			// TODO: this should be removed once we have proper enforcement or error reporting for this
+			clusterSet = pickSingleDestination(clusterSet)
+		}
+
 		// set destinations and enqueue binding for syncing
 		c.bindingPolicyResolver.SetDestinations(bindingPolicy.GetName(), clusterSet)
 		c.logger.V(4).Info("enqueued Binding for syncing, while handling BindingPolicy", "name", bindingPolicy.Name)
 		c.enqueueBinding(bindingPolicy.GetName())
 
 		// requeue objects for re-evaluation
-		if err := c.requeueForBindingPolicyChanges(); err != nil {
-			return err
+		return c.requeueForBindingPolicyChanges()
+	}
+
+	// handle deletion of bindingpolicy
+	return c.handleBindingPolicyDeletion(ctx, bindingPolicy)
+}
+
+func (c *Controller) handleBindingPolicyDeletion(ctx context.Context, bindingPolicy *v1alpha1.BindingPolicy) error {
+	if c.bindingPolicyResolver.ResolutionRequiresSingletonReportedState(bindingPolicy.GetName()) {
+		// if the bindingpolicy required a singleton status, all selected objects should
+		// be requeued in order to remove the label
+		if err := c.requeueSelectedWorkloadObjects(bindingPolicy.GetName()); err != nil {
+			return fmt.Errorf("failed to requeue selected workload objects: %w", err)
 		}
-	} else {
-		c.bindingPolicyResolver.DeleteResolution(bindingPolicy.GetName())
 	}
 
+	// we delete finalizer only after cleaning up in the case above
 	if err := c.handleBindingPolicyFinalizer(ctx, bindingPolicy); err != nil {
-		return err
+		return fmt.Errorf("failed to handle finalizer for BindingPolicy %s: %w", bindingPolicy.Name, err)
 	}
 
-	return c.cleanUpObjectsNoLongerMatching(bindingPolicy)
+	c.bindingPolicyResolver.DeleteResolution(bindingPolicy.GetName())
```

Comment:
Note to self: why was this not needed before?

Reply:
It was moved, not added (in this func).